File: platform/cli-go/internal/commands/runs.go
package commands

import (
	"fmt"
	"strings"

	"github.com/elloloop/devpod/platform/cli-go/internal/api"
	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/spf13/cobra"
)

type runEntry struct {
	ID          string `json:"id"`
	Workflow    string `json:"workflow"`
	Status      string `json:"status"`
	Duration    int    `json:"duration"`
	StartedAt   string `json:"startedAt"`
	CompletedAt string `json:"completedAt"`
	Jobs        []jobEntry `json:"jobs"`
}

type jobEntry struct {
	ID       string      `json:"id"`
	Name     string      `json:"name"`
	Status   string      `json:"status"`
	Duration int         `json:"duration"`
	Steps    []stepEntry `json:"steps"`
}

type stepEntry struct {
	Name     string `json:"name"`
	Status   string `json:"status"`
	Duration int    `json:"duration"`
	Log      string `json:"log"`
}

func newRunsCmd() *cobra.Command {
	var statusFilter string
	var limit string

	runsCmd := &cobra.Command{
		Use:   "runs",
		Short: "List and inspect workflow runs",
		RunE: func(cmd *cobra.Command, args []string) error {
			params := ""
			var parts []string
			if statusFilter != "" {
				parts = append(parts, "status="+statusFilter)
			}
			parts = append(parts, "limit="+limit)
			params = strings.Join(parts, "&")

			var data []runEntry
			if err := api.Get("/api/runs?"+params, &data); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			if len(data) == 0 {
				fmt.Println(format.DimText("No runs found."))
				return nil
			}

			var rows [][]string
			for _, run := range data {
				icon := format.StatusIcon(run.Status)
				dur := "\u2014"
				if run.Duration > 0 {
					dur = fmt.Sprintf("%ds", run.Duration/1000)
				}
				started := "\u2014"
				if run.StartedAt != "" {
					started = format.RelativeTime(run.StartedAt)
				}
				idShort := run.ID
				if len(idShort) > 8 {
					idShort = idShort[:8]
				}
				rows = append(rows, []string{
					idShort,
					run.Workflow,
					fmt.Sprintf("%s %s", icon, run.Status),
					dur,
					started,
				})
			}

			fmt.Println(format.Table([]string{"ID", "Workflow", "Status", "Duration", "Started"}, rows))
			return nil
		},
	}

	runsCmd.Flags().StringVarP(&statusFilter, "status", "s", "", "Filter by status")
	runsCmd.Flags().StringVarP(&limit, "limit", "l", "10", "Limit number of results")

	// runs view <id>
	viewCmd := &cobra.Command{
		Use:   "view <id>",
		Short: "View details of a specific run",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			id := args[0]

			var run runEntry
			if err := api.Get("/api/runs/"+id, &run); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			icon := format.StatusIcon(run.Status)

			fmt.Printf("Run %s\n", run.ID)
			fmt.Printf("Workflow: %s\n", run.Workflow)
			fmt.Printf("Status:   %s %s\n", icon, run.Status)
			if run.Duration > 0 {
				fmt.Printf("Duration: %ds\n", run.Duration/1000)
			}
			if run.StartedAt != "" {
				fmt.Printf("Started:  %s\n", format.RelativeTime(run.StartedAt))
			}
			fmt.Println()

			if len(run.Jobs) > 0 {
				fmt.Println("Jobs:")
				for _, job := range run.Jobs {
					jobIcon := format.StatusIcon(job.Status)
					var jobDur string
					if job.Duration > 0 {
						jobDur = fmt.Sprintf(" \u2014 %ds", job.Duration/1000)
					}
					fmt.Printf("  %s %s (%s)%s\n", jobIcon, job.Name, job.Status, format.DimText(jobDur))

					for _, step := range job.Steps {
						stepIcon := format.StatusIcon(step.Status)
						fmt.Printf("    %s %s\n", stepIcon, step.Name)
					}
				}
			}

			return nil
		},
	}

	// runs logs <id>
	logsCmd := &cobra.Command{
		Use:   "logs <id>",
		Short: "Print full logs for a run",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			id := args[0]

			var run runEntry
			if err := api.Get("/api/runs/"+id, &run); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			if len(run.Jobs) > 0 {
				for _, job := range run.Jobs {
					fmt.Printf("\u2500\u2500 %s \u2500\u2500\n", job.Name)
					for _, step := range job.Steps {
						fmt.Printf("\u25b8 %s\n", step.Name)
						if step.Log != "" {
							fmt.Println(format.Indent(step.Log, 2))
						}
					}
					fmt.Println()
				}
			} else {
				// Fallback: fetch raw logs
				var logs string
				if err := api.Get("/api/runs/"+id+"/logs", &logs); err != nil {
					return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
				}
				fmt.Println(logs)
			}

			return nil
		},
	}

	runsCmd.AddCommand(viewCmd)
	runsCmd.AddCommand(logsCmd)

	return runsCmd
}

File: platform/cli/src/commands/runs.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as api from '../api';
import {
  statusIcon,
  statusColor,
  relativeTime,
  duration,
  table,
  indent,
  errorMessage,
} from '../format';

interface Run {
  id: string;
  workflow: string;
  status: string;
  duration?: number;
  startedAt?: string;
  completedAt?: string;
  jobs?: Job[];
}

interface Job {
  id: string;
  name: string;
  status: string;
  duration?: number;
  steps?: Step[];
}

interface Step {
  name: string;
  status: string;
  duration?: number;
  log?: string;
}

export function registerRunsCommand(program: Command): void {
  const runs = program
    .command('runs')
    .description('List and inspect workflow runs');

  // Default action: list runs
  runs
    .option('-s, --status <status>', 'Filter by status')
    .option('-l, --limit <n>', 'Limit number of results', '10')
    .action(async (opts: { status?: string; limit: string }) => {
      try {
        const params = new URLSearchParams();
        if (opts.status) params.set('status', opts.status);
        params.set('limit', opts.limit);

        const data = await api.get<Run[]>(`/api/runs?${params.toString()}`);

        if (!data || data.length === 0) {
          console.log(chalk.dim('No runs found.'));
          return;
        }

        const rows = data.map((run) => {
          const icon = statusIcon(run.status);
          const color = statusColor(run.status);
          const dur = run.duration ? duration(run.duration) : '\u2014';
          const started = run.startedAt
            ? relativeTime(run.startedAt)
            : '\u2014';
          return [
            chalk.cyan(run.id.slice(0, 8)),
            run.workflow,
            `${icon} ${color(run.status)}`,
            dur,
            chalk.dim(started),
          ];
        });

        console.log(
          table(rows, ['ID', 'Workflow', 'Status', 'Duration', 'Started']),
        );
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // View a specific run
  runs
    .command('view')
    .description('View details of a specific run')
    .argument('<id>', 'Run ID')
    .action(async (id: string) => {
      try {
        const run = await api.get<Run>(`/api/runs/${id}`);

        const icon = statusIcon(run.status);
        const color = statusColor(run.status);

        console.log(`${chalk.bold('Run')} ${chalk.cyan(run.id)}`);
        console.log(`Workflow: ${chalk.bold(run.workflow)}`);
        console.log(`Status:   ${icon} ${color(run.status)}`);
        if (run.duration) {
          console.log(`Duration: ${duration(run.duration)}`);
        }
        if (run.startedAt) {
          console.log(`Started:  ${relativeTime(run.startedAt)}`);
        }
        console.log();

        if (run.jobs && run.jobs.length > 0) {
          console.log(chalk.bold('Jobs:'));
          for (const job of run.jobs) {
            const jobIcon = statusIcon(job.status);
            const jobColor = statusColor(job.status);
            const jobDur = job.duration ? ` \u2014 ${duration(job.duration)}` : '';
            console.log(
              `  ${jobIcon} ${job.name} ${jobColor(`(${job.status})`)}${chalk.dim(jobDur)}`,
            );

            if (job.steps && job.steps.length > 0) {
              for (const step of job.steps) {
                const stepIcon = statusIcon(step.status);
                console.log(`    ${stepIcon} ${step.name}`);
              }
            }
          }
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // Show logs for a run
  runs
    .command('logs')
    .description('Print full logs for a run')
    .argument('<id>', 'Run ID')
    .action(async (id: string) => {
      try {
        const run = await api.get<Run>(`/api/runs/${id}`);

        if (run.jobs && run.jobs.length > 0) {
          for (const job of run.jobs) {
            console.log(chalk.bold(`\u2500\u2500 ${job.name} \u2500\u2500`));
            if (job.steps) {
              for (const step of job.steps) {
                console.log(chalk.cyan(`\u25b8 ${step.name}`));
                if (step.log) {
                  console.log(indent(step.log));
                }
              }
            }
            console.log();
          }
        } else {
          // Fallback: fetch raw logs
          const logs = await api.get<string>(`/api/runs/${id}/logs`);
          console.log(logs);
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}
